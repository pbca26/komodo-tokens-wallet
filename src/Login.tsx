import React, {useState, useEffect} from 'react';
import Logo from './Logo';
import TokensLib from './cclib-import';
import Blockchain from './blockchainWrapper';
import {chains} from './constants';
import writeLog from './log';
import dev from './dev';
import {LoginProps} from './types';

const Login: React.FC<LoginProps> = props => {
  const [privKeyInput, setPrivKeyInput] = useState(dev.seed || '');
  const [chainDropdownOpen, setChainDropdownOpen] = useState(false);
  const [chain, setChain] = useState('');

  const dropdownEventHandler = (e: React.MouseEvent<HTMLInputElement>) => {
    e.stopPropagation();

    setChainDropdownOpen(!chainDropdownOpen);
  };

  const setChainEventHandler = (chain: string) => {
    setChain(chain);
    writeLog(`explorer is set to ${chains[chain].explorerApiUrl}`);
    Blockchain.default.setExplorerUrl(chains[chain].explorerApiUrl);
  };

  const clickOutsideHandler = () => {
    setChainDropdownOpen(false);
  };

  const getWifKeyHandler = () => {
    const ccLibVersion = chains[chain].ccLibVersion === 1 ? 'V1' : 'V2';
    const wif = TokensLib[ccLibVersion].keyToWif(privKeyInput);
    const address = TokensLib[ccLibVersion].keyToCCAddress(wif, 'wif', chains[chain].ccIndex);

    props.setKey({
      wif,
      address,
      chain,
    });

    writeLog('login wif', wif);
    writeLog('login address', address);
    writeLog(`chain ${chain}`, chains[chain]);
  };

  const enableDemo = () => {
    setPrivKeyInput('lime lime3');
    setChain(dev.chain);
    props.setIsDemo(true);
  };

  useEffect(() => {
    if (props.isDemo) {
      writeLog(`explorer is set to ${chains[chain].explorerApiUrl}`);
      Blockchain.mock.setExplorerUrl(chains[chain].explorerApiUrl);  
      getWifKeyHandler();
    }
  }, [props.isDemo]);

  useEffect(() => {
    document.addEventListener(
      'click',
      clickOutsideHandler,
      false
    );

    setChainEventHandler(dev.chain);

    return () => {
      document.removeEventListener(
        'click',
        clickOutsideHandler,
        false
      );
    };
  }, []);

  useEffect(() => {
    writeLog('login state', {
      privKeyInput,
      chainDropdownOpen,
      chain
    });
  });

  return(
    <div className="main">
      <Logo />
      <div className="content login-form">
        <h4>Login</h4>
        <p>Enter your seed phrase or WIF key in the form below</p>
        <div className="input-form">
          <div className={`dropdown${chainDropdownOpen ? ' is-active' : ''}`}>
            <div className={`dropdown-trigger${chain ? ' highlight' : ''}`}>
              <button
                className="button"
                onClick={dropdownEventHandler}>
                <span>{chain ? chain : 'Select chain'}</span>
                <span className="icon is-small">
                  <i className="fas fa-angle-down"></i>
                </span>
              </button>
            </div>
            <div
              className="dropdown-menu"
              id="dropdown-menu"
              role="menu">
              <div className="dropdown-content">
                {Object.keys(chains).map(chainItem => chains[chainItem].enabled ? (
                  <a
                    key={`chain-${chainItem}`}
                    className={`dropdown-item${chain === chainItem ? ' disabled' : ''}`}
                    onClick={() => setChainEventHandler(chainItem)}>
                    <span className="dropdown-balance">{chainItem}</span>
                  </a>
                ) : (null))}
              </div>
            </div>
          </div>
          <input
            type="password"
            name="privKeyInput"
            placeholder="Seed or WIF key"
            value={privKeyInput}
            onChange={(e) => setPrivKeyInput(e.target.value)} />
          <button
            type="button"
            onClick={getWifKeyHandler}
            disabled={
              !privKeyInput ||
              !chain
            }>Login</button>
          <button
            type="button"
            className="load-demo-btn"
            onClick={enableDemo}>Load demo</button>
        </div>
      </div>
    </div>
  );
};

export default Login;