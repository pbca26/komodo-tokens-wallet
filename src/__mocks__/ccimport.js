// IMPORTANT: cclib must be compiled with nodejs target
//            otherwise Jest won't be able to load it
import * as ccimp from '../../cryptoconditions-js-tests/pkg/cryptoconditions.js';
window.ccimp = ccimp;