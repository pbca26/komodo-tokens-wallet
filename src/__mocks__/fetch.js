import mockData from './blockchain.json';

const mockFetch = jest.fn((url) => {
  console.log('fetch url', url);

  return Promise.resolve({
    json: () => {
      if (url.indexOf('status?q=getInfo') > -1) {
        return mockData.getInfo;
      } if (url.indexOf('tx/send') > -1) {
        return {txid: '1111111111111111111111111111111111111111111111111111111111111111'};
      } if (url.indexOf('rawtx') > -1) {
        if (url.indexOf('/')) {
          const temp = url.split('/');
          return mockData.rawtx[temp[1]];
        }

        return mockData.rawtx['890a375c18cbc420d3293dbd661f557caf5e4ca43c770cf479bcc9cfe14b9cbd'];
      } else if (url.indexOf('addr/undefined/utxo') > -1 || url.indexOf('addr/RRF6ocq94kLpaH6wPdZjTqQa9mAmT1GToc/utxo') > -1) {
        return mockData.utxo;
      } else if (url === 'tokens') {
        return mockData.tokens;
      } else if (url.indexOf('tokens/balance') > -1) {
        return mockData.balance;
      } else if (url.indexOf('transactions?address=undefined') > -1 || url.indexOf('transactions?address=CaopajuemreFwEAfbywvFi8oFyrJPNkDs1') > -1) {
        return mockData.transactions;
      } else if (url.indexOf('tokens/orderbook') > -1) {
        return mockData.orderbook;
      } else if (url.indexOf('tokens/createtx') > -1) {
        if (url.indexOf('tokens/createtx?pubkey=03256ba44eeb188404b94ae8ed64f1fe6ad89580375830845361e365598efa3ff3&amount=10000') > -1) {
          return mockData.cancelAsk.createtx;
        } else if (url.indexOf('tokens/createtx?pubkey=03256ba44eeb188404b94ae8ed64f1fe6ad89580375830845361e365598efa3ff3&amount=11000' > -1)) {
          return mockData.sellToken.createtx;
        }
        return mockData.createtx;
      } else if (url.indexOf('tokens/addccinputs') > -1) {
        if (url.indexOf('tokens/addccinputs?pubkey=03256ba44eeb188404b94ae8ed64f1fe6ad89580375830845361e365598efa3ff3&tokenid=7e2b623cb57b44b8dc5d3a3cc36c20d08f66023a85fa2a4490a8fb8783d38347&amount=1' > -1)) {
          return mockData.sellToken.addccinputs;
        }
        return mockData.createToken.fixedSupplyToken.chainData;
      } else if (url.indexOf('tokens/utxo') > -1) {
        return mockData.ccUtxo;
      } else if (url.indexOf('tokens/transactionsmany?txid1') > -1) {
        if (url.indexOf('tokens/transactionsmany?txid1=b8ceac7c2130b9a44fd11a1b67f4fa7b7a908b9f9a58e4f0024acd3b486b7a90&txid2=572663bb71d9822cf70d60ed4d180300f61bf8790dc5c2023d1888f1b8e8c254') > -1) {
          return mockData.cancelAsk.transactionsmany;
        }
        return mockData.transactionsMany;
      }
    },
    headers: {
      get: () => {
        return 'application/json';
      },
    },
    ok: true,
  });
});

export default mockFetch;