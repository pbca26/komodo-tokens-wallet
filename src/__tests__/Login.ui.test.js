import React from 'react';
import renderer from 'react-test-renderer';
import {render, fireEvent, screen} from '@testing-library/react';
import Login from '../Login';
import '../__mocks__/ccimport';

// TODO: test keying input data actually changes state

test('Login component render check', () => {
  const component = renderer.create(
    <Login />,
  );
  
  let treeJSON = component.toJSON();
  expect(treeJSON).toMatchSnapshot();

  let tree = component.toTree();

  const mProps = {
    setKey: function(obj) {
      expect(obj).toEqual({
        wif: 'UtaKVKuNvDVYxqS8PhRESwKeRYHhzYYkEGgvwEJ9XvghuRVs1VNM',
        address: {
          normal: 'RRF6ocq94kLpaH6wPdZjTqQa9mAmT1GToc',
          cc: 'CaopajuemreFwEAfbywvFi8oFyrJPNkDs1',
          pubkey: '03256ba44eeb188404b94ae8ed64f1fe6ad89580375830845361e365598efa3ff3'
        },
        chain: 'TKLTEST'
      });
    }
  };
  const {getByText, getAllByText, getByPlaceholderText} = render(<Login {...mProps} />);

  // check if all labels/titles/form elements are present
  expect(getAllByText('Login')[0]).toBeDefined();
  expect(getByText('Enter your seed phrase or WIF key in the form below')).toBeDefined();
  expect(getByPlaceholderText('Seed or WIF key')).toBeDefined();
  // check if chain is set to TKLTEST
  expect(tree.instance.state.chain).toBe('TKLTEST');
  expect(getAllByText('TKLTEST')[0]).toBeDefined();
  expect(getAllByText('TKLTEST')[1]).toBeDefined();

  fireEvent.change(getByPlaceholderText('Seed or WIF key'), {target: {value: 'lime lime3'}});
  // trigger login event
  fireEvent.click(screen.getAllByText(/Login/i)[1]);
});