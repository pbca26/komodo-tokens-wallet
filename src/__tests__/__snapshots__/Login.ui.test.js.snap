// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component render check 1`] = `
<div
  className="main"
>
  <div
    className="app-logo"
  >
    <div
      className="box"
    />
    <div
      className="circle"
    />
    <img
      src="https://explorer.komodoplatform.com/public/img/coins/kmd.png"
    />
  </div>
  <div
    className="content login-form"
  >
    <h4>
      Login
    </h4>
    <p>
      Enter your seed phrase or WIF key in the form below
    </p>
    <div
      className="input-form"
    >
      <div
        className="dropdown"
      >
        <div
          className="dropdown-trigger highlight"
        >
          <button
            className="button"
            onClick={[Function]}
          >
            <span>
              TKLTEST
            </span>
            <span
              className="icon is-small"
            >
              <i
                className="fas fa-angle-down"
              />
            </span>
          </button>
        </div>
        <div
          className="dropdown-menu"
          id="dropdown-menu"
          role="menu"
        >
          <div
            className="dropdown-content"
          >
            <a
              className="dropdown-item"
              onClick={[Function]}
            >
              <span
                className="dropdown-balance"
              >
                TOKENSV2
              </span>
            </a>
            <a
              className="dropdown-item"
              onClick={[Function]}
            >
              <span
                className="dropdown-balance"
              >
                TKLTEST
              </span>
            </a>
          </div>
        </div>
      </div>
      <input
        name="privKeyInput"
        onChange={[Function]}
        placeholder="Seed or WIF key"
        type="password"
        value="lime lime3"
      />
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Login
      </button>
    </div>
  </div>
</div>
`;
